iree/samples/static_library/static_library_demo.c

```diff
@@ -0,0 +1,214 @@
+// released by the caller.
+iree_status_t create_device_with_static_loader(iree_hal_device_t** device) {
+  iree_status_t status = iree_ok_status();
+  // Set paramters for the device created in the next step.
+  iree_hal_task_device_params_t params;
+  iree_hal_task_device_params_initialize(&params);
+
+  // Load the statically embedded library
+  const iree_hal_executable_library_header_t** static_library =
+      simple_mul_dispatch_0_library_query(
+          IREE_HAL_EXECUTABLE_LIBRARY_LATEST_VERSION, /*reserved=*/NULL);
+  const iree_hal_executable_library_header_t** libraries[1] = {static_library};
+
+  iree_hal_executable_loader_t* library_loader = NULL;
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_static_library_loader_create(
+        IREE_ARRAYSIZE(libraries), libraries,
+        iree_hal_executable_import_provider_null(), iree_allocator_system(),
+        &library_loader);
+  }
+
+  iree_task_executor_t* executor = NULL;
+  if (iree_status_is_ok(status)) {
+    iree_task_executor_create_from_flags(iree_allocator_system(), &executor);
+  }
+
+  // Create the device and release the executor and loader afterwards.
+  if (iree_status_is_ok(status)) {
+    iree_hal_task_device_create(iree_make_cstring_view("dylib"), &params,
+                                executor, 1, &library_loader,
+                                iree_allocator_system(), device);
+  }
+  iree_task_executor_release(executor);
+  iree_hal_executable_loader_release(library_loader);
+
+  return status;
+}
+
+iree_status_t Run() {
+  iree_status_t status = iree_ok_status();
+
+  // Instance configuration (this should be shared across sessions).
+  iree_runtime_instance_options_t instance_options;
+  iree_runtime_instance_options_initialize(IREE_API_VERSION_LATEST,
+                                           &instance_options);
+  iree_runtime_instance_options_use_all_available_drivers(&instance_options);
+  iree_runtime_instance_t* instance = NULL;
+
+  if (iree_status_is_ok(status)) {
+    status = iree_runtime_instance_create(&instance_options,
+                                          iree_allocator_system(), &instance);
+  }
+
+  // Create dylib device with static loader.
+  iree_hal_device_t* device = NULL;
+  if (iree_status_is_ok(status)) {
+    status = create_device_with_static_loader(&device);
+  }
+  iree_vm_module_t* hal_module = NULL;
+  if (iree_status_is_ok(status)) {
+    status =
+        iree_hal_module_create(device, iree_allocator_system(), &hal_module);
+  }
+
+  // Session configuration (one per loaded module to hold module state).
+  iree_runtime_session_options_t session_options;
+  iree_runtime_session_options_initialize(&session_options);
+  iree_runtime_session_t* session = NULL;
+  if (iree_status_is_ok(status)) {
+    status = iree_runtime_session_create_with_device(
+        instance, &session_options, device,
+        iree_runtime_instance_host_allocator(instance), &session);
+  }
+  iree_hal_device_release(device);
```

Comment:
you release the device and then use it down below - that's no good (you can move this down to the very bottom of the function)